url146c819f25_jlink_Parse4SJM_tgz-pJ8-sjm_examples_book_robot_ShowRobotRefactoredJ8/src/sjm/grammar/Grammar.java

```diff
@@ -0,0 +1,222 @@
+		}
+	}
+
+	private Set<IParser> allParsersUsedFromStartRule() {
+		Set<IParser> all = ParserCollector.collectAllReferencedParsers(startParser());
+		all.add(startParser());
+		return all;
+	}
+
+	private void checkGrammarHasStartRule() {
+		if (startParser() == null)
+			throw new GrammarException("A start rule must be specified");
+	}
+
+	public IParser startParser() {
+		return rules.get(startRule);
+	}
+
+	public IParser getRule(String ruleName) {
+		return rules.get(ruleName);
+	}
+
+	public String defineRule(String ruleText) {
+		return (String) getRuleGrammar().parse(ruleText).getStack().pop();
+	}
+
+	public String defineRule(String ruleText, IParserMatched matched) {
+		String ruleName = defineRule(ruleText);
+		addAssembler(ruleName, new ParserMatchedAssembler(matched));
+		return ruleName;
+	}
+
+	public String defineRule(String ruleText, final Closure closure) {
+		IParserMatched matched = new IParserMatched() {
+			public void apply(List<Object> tokens, Stack<Object> stack) {
+				closure.call(new Object[] { tokens, stack });
+			}
+		};
+		return defineRule(ruleText, matched);
+	}
+
+	/**
+	 * Create lazily since not all Grammars need it
+	 */
+	private Grammar getRuleGrammar() {
+		if (ruleGrammar == null)
+			ruleGrammar = new RuleGrammar(this);
+		return ruleGrammar;
+	}
+
+	public void addAssembler(String ruleName, IAssembler assembler) {
+		if (!ruleExists(ruleName)) {
+			throw new GrammarException("Rule '" + ruleName + "' does not exist.");
+		}
+		getRule(ruleName).setAssembler(assembler);
+	}
+
+	private boolean ruleExists(String ruleName) {
+		return rules.containsKey(ruleName);
+	}
+
+	public void initTokenizer(InitTokenizer initializer) {
+		this.initTokenizer = initializer;
+	}
+
+	public void printOn(PrintWriter writer) {
+		printRuleOn(startRule, writer);
+		for (String ruleName : rules.keySet()) {
+			if (ruleName.equals(startRule))
+				continue;
+			printRuleOn(ruleName, writer);
+		}
+
+	}
+
+	private void printRuleOn(String ruleName, PrintWriter writer) {
+		String ruleString = getRule(ruleName).toString();
+		if (ruleString.startsWith("(") && ruleString.endsWith(")"))
+			ruleString = ruleString.substring(1, ruleString.length() - 1);
+		writer.println(ruleName + " = " + ruleString + ";");
+	}
+
+	public void registerTerminal(Class<? extends Terminal> terminalClass) {
+		String terminalType = terminalClass.getSimpleName();
+		registerTerminal(terminalType, terminalClass);
+	}
+
+	public void registerTerminal(String terminalType, Class<? extends Terminal> terminalClass) {
+		terminalTypes.put(terminalType.toLowerCase(), terminalClass);
+	}
+
+	public Terminal terminal(String terminalType) {
+		Class<? extends Terminal> terminalClass = terminalTypes.get(terminalType.toLowerCase());
+		if (terminalClass == null)
+			throw new GrammarException("Unknown terminal type: " + terminalType);
+		try {
+			return terminalClass.newInstance();
+		} catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)